### DISCUSSION TITLE
How can I validate a nested object onBlur

### DISCUSSION THREAD

**Discussion by engelkes-finstreet** [react-hook-form/react-hook-form#4174] [Q&A]
Currently I am trying to get validation with Zod for my PasswordForm working - this is just a contrived example to easier show my problem.
Here is the Codesandbox https://codesandbox.io/s/react-hook-form-typescript-forked-755ri?file=/src/index.tsx
Let me explain my problem. I have mode: "onBlur" and the following validations

```js
const LoginInput = z.object({
  userName: z.string().min(5),
  passwordForm: z
    .object({
      password: z.string().min(3),
      confirm: z.string().min(3)
    })
    .refine(
      (data) => {
        return data.password === data.confirm;
      },
      {
        message: "Passwords do not match"
      }
    )
});
```

The userName validation is working as expected and instantly shows on blur. But the validation for the nested object seems not to work on blur. Inside the resolver the errors get logged correctly but the context seems not to update. Only after hitting the submit button all the errors for the nested object - either the length of the fields itself or the not matching password - will be available via context.

Is there anything I am doing wrong? How can I validate this nested object `onBlur`?

I tried this with the current `6.15.1` and the `7.0.0-beta.3`version and with the @hooksform zodResolver and this current custom resolver which you see in the Codesandbox.

**Comment by bluebill1049** [comment]
Sounds like dependant field validation, in hook form we focus on single input at a time, if you need to trigger other inputs' validation, you will need to use `triggger` api.

**Reply by engelkes-finstreet** [reply]
When should I call the `trigger` api? I added an `onChange` method to my `password` and `confirm` field. I do now get the validations in context BUT it is "lacking" behind.
Entering `123` as password and than `123` as confirm it still shows that `Passwords do not match`. After entering another character to confirm the validation shows as correct

**Reply by bluebill1049** [reply]
hope this helsp: https://codesandbox.io/s/react-hook-form-typescript-forked-c65jc?file=/src/index.tsx

**Comment by jason-nivelo** [comment]
A little bit of a cleaner solution IMO, assuming you are disabling the form submit button until `isValid` returns true

```
const signUpFormSchema = z.object({
  ...
  password: z.string().min(10, { message: 'Password must be 10 characters' }),
  confirm: z.string().min(1, { message: 'Required' }),
});

const checkConfirmPassword = () => {
  const { password, confirm } = getValues();
  if (password !== confirm) {
    setError('confirm', { message: 'Password do not match' });
  }
};

  <Input
    label='Password'
    id='password'
    type='password'
  />
  <Input
    label='Confirm Password'
    id='confirm'
    type='password'
    onBlur={checkConfirmPassword}
  />

```

**Comment by Peroconino** [comment]
Having a similar problem, but its typescript that is actually complaining about something:
So basically i have my nested schema, and i want to validate them using react-hook-form.

validation.ts
```ts
const LoginInput = z.object({
  userName: z.string().min(5),
  passwordForm: z
    .object({
      password: z.string().min(3),
      confirm: z.string().min(3)
    })
    .refine(
      (data) => {
        return data.password === data.confirm;
      },
      {
        message: "Passwords do not match"
      }
    )
});

export type formSchemaType = Required<
  z.infer<typeof formSchema>
>;

export const formResolver = zodResolver(
  formSchema
);
```

component.ts
```ts
const {
    register,
    handleSubmit,
    control,
    formState: { errors },
  } = useForm<formSchemaType >({
    resolver: formResolver,
    mode: "onChange",
    delayError: 1000,
  });

return (
   <Input
         name="passwordForm.password" // <----- typescript says it does not exist
         {...register(name, options)}
      />
);
```
The typing of the possible names are the **keyof formSchemaType**, so it is not fitting anymore when it has nested objects, does someone know how to type it correctly for nested objects?

EDIT: heres a example in [codesandbox](https://codesandbox.io/s/objective-napier-qxv4z4?file=/src/App.tsx)

**Comment by Peroconino** [comment]
Just figured a way out. Basically i had to find a way to create a type that extracts the keys and nested keys.
```ts
type NestedKey<O extends Record<string, unknown>> = {
  [K in Extract<keyof O, string>]: O[K] extends Array<any>
    ? K
    : O[K] extends Record<string, unknown>
    ? `${K}` | `${K}.${NestedKey<O[K]>}`
    : K;
}[Extract<keyof O, string>];

type nestedKeys = NestedKey<formSchemaType>;
```
That type will be able to give me the schema typing correctly.

**Reply by ohmyjersh** [reply]
@LeandroFontellas do you have an example (codesandbox) of using that NestedKey type where you're able to get the validation working on nested objects?

**Reply by Peroconino** [reply]
@ohmyjersh you can see it working in this [TS Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAchDOwIBMDSEQB4DyUIA8kA7ZeKAJQgGMB7AJ2U0ToEsiBzAGigFciBrIjQDuRAHxioAXigBvALAAoKFADaqKGygBRQnQCGVYJn4YaAMyjZuzNuzEBdAFxX1DvIQgkyAQToGsfSIQMSUVFQB+KFQw8JdsNw9iUgpqekZbDm4+QRFxWMioAAMAEllUAF8iqAAfYrLKgDoyuEQUdCwE1AcxKoKoF1QAbiUK1V1gAyMTM0trKEz7BxHFJVoiRChIOngaImk5fqJ9AFsIFwByMHMWdh4IFjoUGgvOfv12c6gAZgAGN+U4WQNHYLgUgPC4TyEDoYP6kMhxzOl2ut3uj2eF3h4Qq-VxigqKyUyGoABt9E8oOY+EYWHsFvRgAAhLAAFSSXhSNAARgArajAMQACm2uyITlZNkZdh8wGALlaSDQGEwrLEAEoVrs6MyQCKYWLuAAiYHsRrQuiNJEQI3qoA)

**Reply by a1eks3y** [reply]
Update: You can use FieldPath<z.infer<typeof formSchema>>
